thetis/interpolation.py

```diff
@@ -0,0 +1,634 @@
+        temp = np.take(tri.transform, simplex, axis=0)
+        delta = ntarget_xyz - temp[:, d]
+        bary = np.einsum('njk,nk->nj', temp[:, :d, :], delta)
+        self.vtx = vertices
+        self.wts = np.hstack((bary, 1 - bary.sum(axis=1, keepdims=True)))
+        self.outside = np.nonzero(np.any(self.wts < 0, axis=1))[0]
+        self.fill_nearest = self.fill_mode == 'nearest' and len(self.outside) > 0
+        if self.fill_nearest:
+            # find nearest neighbor in the data set
+            from scipy.spatial import cKDTree
+            dist, ix = cKDTree(ngrid_xyz).query(ntarget_xyz[self.outside])
+            self.outside_to_nearest = ix
+
+    def _normalize_coords(self, xyz):
+        if self.normalize:
+            return self.norm_a*xyz + self.norm_b
+        return xyz
+
+    def __call__(self, values):
+        """
+        Interpolate values defined on grid_xyz to target_xyz.
+
+        :arg values: Array of source values to interpolate, shape (npoints, )
+        :kwarg float fill_value: Fill value to use outside the source grid (default: NaN)
+        """
+        ret = np.einsum('nj,nj->n', np.take(values, self.vtx), self.wts)
+        if self.fill_mode is None:
+            ret[self.outside] = self.fill_value
+        elif self.fill_nearest:
+            ret[self.outside] = values[self.outside_to_nearest]
+        return ret
+
+
+class FileTreeReader(object):
+    """
+    Abstract base class of file tree reader object
+    """
+    @abstractmethod
+    def __call__(self, filename, time_index):
+        """
+        Reads a data for one time step from the file
+
+        :arg str filename: a filename where to find the data (e.g. filename)
+        :arg int time_index: time index to read
+        :return: a list of floats or numpy.array_like objects
+        """
+        pass
+
+
+class NetCDFTimeSeriesReader(FileTreeReader):
+    """
+    A simple netCDF reader that returns a time slice of the given variable.
+
+    This class does not interpolate the data in any way. Useful for
+    interpolating time series.
+    """
+    def __init__(self, variable_list, time_variable_name='time'):
+        self.variable_list = variable_list
+        self.time_variable_name = time_variable_name
+        self.time_dim = None
+        self.ndims = None
+
+    def _detect_time_dim(self, ncfile):
+        assert self.time_variable_name in ncfile.dimensions
+        nc_var = ncfile[self.variable_list[0]]
+        assert self.time_variable_name in nc_var.dimensions
+        self.time_dim = nc_var.dimensions.index(self.time_variable_name)
+        self.ndims = len(nc_var.dimensions)
+
+    def _get_slice(self, time_index):
+        """
+        Returns a slice object that extracts a single time index
+        """
+        if self.ndims == 1:
+            return time_index
+        slice_list = [slice(None, None, None)]*self.ndims
+        slice_list[self.time_dim] = slice(time_index, time_index+1, None)
+        return slice
```

Comment:
This looks wrong.  Did you mean to `return slice_list`?

Reply:
oops, fixed.